mitsurunner.yaml

```diff
@@ -81,96 +90,172 @@ sensor:
+                }
+                break;
+
+            case ST_START_DEFROSTING:
+                if (temperature_delta < 0.0) {
+                    id(EnterDefrostingStarted).execute();
+                }
+                // Timeout to go back to the idle state, if defrosting has not been started
+                else if (id(G_state_time_passed)) {
+                    id(EnterIdle).execute();
+                }
+                break;
+
+            case ST_DEFROSTING_STARTED:
+                if (id(G_state_time_passed)) {
+                    id(EnterHeatingMinTime).execute();
+                }
+                break;
+
+            case ST_HEATING_MIN_TIME:
+                if (id(G_state_time_passed)) {
+                    id(EnterIdle).execute();
+                }
+                break;
+        }
         return temperature_delta;
 
 
-# defrosting script
 script:
-  - id: trigger_defrosting
-    mode: single #will be started only if there were no ongoing defrosting already
-    then:    
-    - switch.turn_off: gpio_relay    
-    - script.execute: start_min_heating_time_timer
-    - script.execute: schedule_forced_defrosting
+
+# -------- State machine state entrance scripts --------
+
+  - id: EnterIdle
+    mode: single
+    then:
+    #Stop other state scripts
+    - script.stop: EnterTempExceeded
+    - script.stop: EnterStartDefrosting
+    - script.stop: EnterDefrostingStarted
+    - script.stop: EnterHeatingMinTime
+    - switch.turn_on: gpio_relay
+
     - lambda: !lambda |-
-        id(mqtt_client).publish(TOPIC_DEFROST_HACK_STATE, to_string(0.0));
-    - delay: 30min    
+        id(G_state) = ST_IDLE;
+        id(G_state_time_passed) = false;
+
+
+  - id: EnterTempExceeded
+    mode: single
+    then:
+    #Stop other state scripts
+    - script.stop: EnterIdle
+    - script.stop: EnterStartDefrosting
+    - script.stop: EnterDefrostingStarted
+    - script.stop: EnterHeatingMinTime
     - switch.turn_on: gpio_relay
+
     - lambda: !lambda |-
-        id(mqtt_client).publish(TOPIC_DEFROST_HACK_STATE, to_string(1.0));
+        id(G_state) = ST_TEMP_EXCEEDED;
+        id(G_state_time_passed) = false;
+
+    - delay: 5min
+    - lambda: !lambda |-
+        id(G_state_time_passed) = true;
 
 
+  - id: EnterStartDefrosting
+    mode: single
+    then:
+    #Stop other state scripts
+    - script.stop: EnterIdle
+    - script.stop: EnterTempExceeded
+    - script.stop: EnterDefrostingStarted
+    - script.stop: EnterHeatingMinTime
+    - switch.turn_off: gpio_relay
 
-# scipt implementing timer for minimum heating time
-# G_min_heating_time_passed is used through this timer to prevent defrosting too quickly after previous defrosting
-  - id: start_min_heating_time_timer
-    mode: single #will be started only if there were no ongoing defrosting already
-    then:  
     - lambda: !lambda |-
-        id(G_min_heating_time_passed) = false;
-        id(mqtt_client).publish(TOPIC_IS_HEATING_MIN_TIME_PASSED, to_string(0.0));
```

Comment:
These MQTT topics is used to do automatic testing of algorithms. 
The logic can be seen with the temperatures + temperature delta graphs together. So for version 1.0 which is directed to hackers, let's keep these untill we are sure that algorithm works.

Reply:
Good point, I'll restore this to the right place.